provider/datagen/src/lib.rs

```diff
@@ -73,86 +72,137 @@ mod testutil;
 mod transform;
 
 pub use error::{is_missing_cldr_error, is_missing_icuexport_error};
-pub use registry::*;
-pub use source::{CollationHanDatabase, CoverageLevel, SourceData};
+pub use registry::{all_keys, all_keys_with_experimental, deserialize_and_discard};
+pub use source::SourceData;
 
-#[allow(clippy::exhaustive_enums)] // exists for backwards compatibility
-#[doc(hidden)]
-#[derive(Debug)]
-pub enum CldrLocaleSubset {
-    Ignored,
-}
-
-impl Default for CldrLocaleSubset {
-    fn default() -> Self {
-        Self::Ignored
-    }
-}
-
-impl CldrLocaleSubset {
-    #[allow(non_upper_case_globals)]
-    pub const Full: Self = Self::Ignored;
-    #[allow(non_upper_case_globals)]
-    pub const Modern: Self = Self::Ignored;
-}
-
-/// [Out::Fs] serialization formats.
-pub mod syntax {
-    #[doc(no_inline)]
-    pub use icu_provider_fs::export::serializers::bincode::Serializer as Bincode;
-    #[doc(no_inline)]
-    pub use icu_provider_fs::export::serializers::json::Serializer as Json;
-    #[doc(no_inline)]
-    pub use icu_provider_fs::export::serializers::postcard::Serializer as Postcard;
-}
+#[cfg(feature = "provider_baked")]
+pub mod baked_exporter;
+pub mod options;
 
 /// A prelude for using the datagen API
 pub mod prelude {
-    #[doc(hidden)]
-    pub use crate::CldrLocaleSubset;
     #[doc(no_inline)]
-    pub use crate::{syntax, BakedOptions, CollationHanDatabase, CoverageLevel, Out, SourceData};
+    pub use crate::{options, DatagenProvider, SourceData};
     #[doc(no_inline)]
     pub use icu_locid::{langid, LanguageIdentifier};
     #[doc(no_inline)]
-    pub use icu_provider::KeyedDataMarker;
+    pub use icu_provider::{datagen::DataExporter, DataKey, KeyedDataMarker};
+
+    // SEMVER GRAVEYARD
+    #[cfg(feature = "legacy_api")]
+    #[doc(hidden)]
+    pub use crate::options::{CollationHanDatabase, CoverageLevel};
+    #[cfg(feature = "legacy_api")]
+    #[allow(deprecated)]
+    #[doc(hidden)]
+    pub use crate::{syntax, BakedOptions, CldrLocaleSubset, Out};
 }
 
 use icu_provider::datagen::*;
 use icu_provider::prelude::*;
-use icu_provider_adapters::empty::EmptyDataProvider;
-use icu_provider_adapters::filter::Filterable;
-use icu_provider_fs::export::serializers::AbstractSerializer;
-use prelude::*;
-use rayon::prelude::*;
 use std::collections::HashSet;
-use std::io::{BufRead, BufReader};
-use std::path::{Path, PathBuf};
+use std::path::Path;
 
 /// [`DataProvider`] backed by [`SourceData`]
-#[allow(clippy::exhaustive_structs)] // any information will be added to SourceData
 #[derive(Debug, Clone)]
+#[cfg_attr(feature = "networking", derive(Default))]
+#[cfg_attr(not(doc), allow(clippy::exhaustive_structs))]
+#[cfg_attr(doc, non_exhaustive)]
```

Comment:
Question: why the special case for docs? If this is a non exhaustive struct, it should be written as such in code. We can always document something as being non exhaustive, but the only purpose of the non_exhaustive annotation is to actually enforce that.

Reply:
It should be an exhaustive struct, but we cannot make it exhaustive because that would be semver breaking (it's currently possible to struct-construct this). We can make it *look* like an exhaustive struct in Rustdoc while not breaking existing code.

We could also mark the field `#[doc(hidden)]`, then the docs would show `/* private fields */` but not `#[non_exhaustive]`.